src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java

```diff
@@ -26,50 +28,106 @@
-                return true;
+                try {
+                    setDisplayName(build, listener);
+	            return true;                   
+                } catch (IllegalArgumentException e){
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
-        };
+         };
     }
-
-    private void setDisplayName(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
-        try {
-            build.setDisplayName(TokenMacro.expandAll(build, listener, template));
-        } catch (MacroEvaluationException e) {
+    private void setDisplayName(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException, IllegalArgumentException {
+    	try{
+            switch(scenario) {
+            case 0:         
+                build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+                return;
+            case 1:
+                if(started) {	
+                    build.setDisplayName(TokenMacro.expandAll(build, listener, matrixTemplate));
+                }else {
+                    build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+                }
+                return;
+            case 2:
+                build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+                return;
+            case -1:
+                listener.getLogger().println("No valid values set for build names!");
+            }
+    	} catch (MacroEvaluationException e) {
             listener.getLogger().println(e.getMessage());
-        }
+    	}  
     }
 
     public MatrixAggregator createAggregator(MatrixBuild build, Launcher launcher, BuildListener listener) {
         return new MatrixAggregator(build,launcher,listener) {
             @Override
             public boolean startBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.startBuild();
+                try {
+                    setDisplayName(build,listener);
+                    started = true;
+	            return super.startBuild();
+                } catch (IllegalArgumentException e){
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
 
             @Override
             public boolean endBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.endBuild();
+                try{
+                    setDisplayName(build,listener);
+                    return super.endBuild();
+                } catch (IllegalArgumentException e){
+                    listener.getLogger().println(e.getMessage());   
+                }
+                return false;
             }
         };
     }
+   
 
-    @Extension
+    @Extension(optional = true)
```

Comment:
Yes `Extension.optional` exists and yes it sometimes works, but you have to know exactly what you are doing. Basically all it does is tell Jenkins that if _when loading the extension_ a `LinkageError` (typically `NoClassDefFoundError`) is thrown, to quietly skip the extension rather than print a big stack trace. For some simple cases, like extensions that just implement an SPI in some other plugin, that is enough. But making the dependency optional here is much more complicated than that.